### PR TITLE
feat#741: AI-instructed Search-Filter Matching

### LINKED ISSUE #741: Enhance Search-Filters with AI-instructed Matching
Enhance Search-Filter with an extra text-field allowing to precisely describe the desired matches.
Then on every hit run an LLM against it.

Yields maximum matching-confidence.

Even better: Provide user with reasoning on why we matched it.

### PR DESCRIPTION
**Pull request by bruderj15**
Implements AI-enhanced search filter matching after percolation, using the `enhanced_search_description` field on `UserSearchFilter` and `UserSearchFilterSummary`.

### Extended `string_newtype!` Macro
- Added optional `max_length(N)` variant to the `string_newtype!` macro: `string_newtype!(Name, max_length(N))` trims whitespace and truncates input to N bytes at a safe boundary, following the same `split_at_checked` pattern used by `Description`.
- Applied to `EnhancedSearchDescription` with a 500-character limit.

### New `EnhancedSearchMatchService`
- Created `EnhancedSearchMatchService` trait (with `#[mockall::automock]`) and `EnhancedSearchMatchServiceImpl` using Google Gemini 2.5 Flash Lite via the `llm` crate.
- The service accepts domain types (`&EnhancedSearchDescription`, `&Title`, `&Description`) and the user's preferred language, determines whether the product matches (binary decision), and returns a compact, user-facing reason only when the product matches.
- Resilient LLM configuration: `temperature(0.0)`, `max_tokens(256)`, `timeout_seconds(30)`, `resilient(true)` with 3 retry attempts, 500ms–5s exponential backoff with jitter.

### New `EnhancedMatchReason` Domain Type
- Created `EnhancedMatchReason` using the `string_newtype!` macro in the `search-filter` core module.

### Updated `SearchFilterProductMatch` and Related Types
- Added `enhanced_match_reason: Option<EnhancedMatchReason>` to `SearchFilterProductMatch` (domain type).
- Added `enhanced_match_reason: Option<String>` to `UserSearchFilterMatchRecord` (DynamoDB record) with `serde(default, skip_serializing_if)`.

### Updated `UserSearchFilterSummary`
- Added `enhanced_search_description: Option<EnhancedSearchDescription>` to `UserSearchFilterSummary` so it is carried through from OpenSearch percolation results, enabling post-percolation AI matching.

### Percolate Lambda Integration
- Wired `EnhancedSearchMatchService` and `UserService` into `ProductEventSearchFilterNotificationsServiceImpl::determine_notification_commands`.
- After percolation, filters with `enhanced_search_description` are evaluated by the LLM: confirmed matches get the reason attached, rejected matches are filtered out, and LLM errors gracefully fall through without a reason.
- User's preferred language is resolved via `UserService::find_user`, defaulting to English on error.
- Created `NotificationCommandWithMatchReason` wrapper to propagate the reason through the notification flow to `SearchFilterProductMatch` creation.
- Updated `main.rs` to wire `GEMINI_API_KEY` env var and new service dependencies.

### CloudFormation Templates
- Added `GEMINI_API_KEY` environment variable to `SearchFilterPercolateProductLambda` in all three templates: ephemeral (`"test-key"`), dev (`{{resolve:ssm:/secrets/dev/gemini-api-key}}`), and prod (`{{resolve:ssm:/secrets/prod/gemini-api-key}}`).

### Dependencies
- Added `llm` (with `google` and `rustls-tls` features) as an optional dependency to the `search-filter` crate, gated behind the `service` feature flag.

## Testing

- Comprehensive unit tests for the `string_newtype!` macro: plain and bounded variants covering truncation, trimming, exact boundary, empty input, display, and conversion.
- Unit tests for `EnhancedSearchDescription` max-length behavior (truncation, trimming, boundary cases).
- Comprehensive unit tests for LLM response parsing (valid/invalid, case-insensitive, multi-language, extra content handling, reason-only-on-match semantics).
- Tests for enhanced match integration: confirmation with reason, rejection (empty result), LLM error fallback, user-not-found language default, and mixed enhanced+plain filter scenarios.
- All existing tests updated for the new fields and passing.

### CODE REVIEW

**Comment by bruderj15** [comment]
@copilot+claude-opus-4.6 

In the previous two commits I added a field `enhanced_search_description` to `UserSearchFilter` & `UserSearchFilterSummary`.
I now want you to implement the matching after percolation to respect that field.

In fact I want you to:
- create a new service (via trait) around `LLM` from crate `llm` (check other usages in the repository for design/structure) which 
  - given the enhanced-search-description (raw user-input explaining in detail what they want, e.g. `"Golden cufflinks with 800 Silver and 24K Gold plating, perfect condition, hammered style with a red rubin in the centre. The rubin must be real and not synthetic..."`) and the product-title and -description, the LLM must determine whether the enhanced description from the user actually matches that product. This is a binary decision. 
  - Next to the decision I want a `"reason"` in the response, where the LLM SHORTLY/COMPACTLY (this is user-facing!!!) explains the decision. The `"reason"` text must be in the users preferred language (load it via `UserService::find_user`)
  - Use a single prompt here which answers with binary decision and reason in one go
- The reason must be attached to `SearchFilterProductMatch` and related types (DTOs) - add it as optional field. Use `string_newtype!(..)` macro for the domain-type, simple `String` for the DTOs.
- unit-test with full coverage

File: src/search-filter/src/data/search_filter_product_match_data.rs [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+use common::product_id::ProductId;
+use common::shop_id::ShopId;
+use common::shops_product_id::ShopsProductId;
+use common::user_id::UserId;
+use serde::{Deserialize, Serialize};
+use time::OffsetDateTime;
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct SearchFilterProductMatchData {
```
Comment (bruderj15) [inline src/search-filter/src/data/search_filter_product_match_data.rs]: I don't think we need this type/file at all. Remove it.

**Review by bruderj15 — COMMENTED** [review]
@copilot great start.

Now:
- intruct the LLM Service to ONLY respond with the reason when `matches` is true.
- wire the new service in the percolate-lambda (new service usage belongs into `ProductEventSearchFilterNotificationsServiceImpl::determine_notification_commands`)
- add tests accordingly

File: src/search-filter-lambda/src/search-filter-lambda-percolate-product/src/main.rs [hunk 2]
```diff
@@ -40,6 +43,8 @@ async fn main() -> Result<(), Error> {
         std::env::var("STAGE_NAME").expect("shouldn't fail loading env-var 'STAGE_NAME'");
     let commit_sha =
         std::env::var("COMMIT_SHA").expect("shouldn't fail loading env-var 'COMMIT_SHA'");
+    let gemini_api_key =
+        std::env::var("GEMINI_API_KEY").expect("shouldn't fail loading env-var 'GEMINI_API_KEY'");
```
Comment (bruderj15) [inline src/search-filter-lambda/src/search-filter-lambda-percolate-product/src/main.rs:47]: make sure to adjust all cloudformation templates in dir `cfn` accordingly to this new required env-var!

File: src/search-filter/src/service/enhanced_search_match_service.rs [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+}
+
+#[async_trait::async_trait]
+#[mockall::automock]
+pub trait EnhancedSearchMatchService {
+    async fn evaluate(
+        &self,
+        enhanced_search_description: &str,
+        product_title: &str,
+        product_description: &str,
```
Comment (bruderj15) [inline src/search-filter/src/service/enhanced_search_match_service.rs]: Keep using the domain-types here.

```suggestion
        enhanced_search_description: &EnhancedSearchDescription,
        product_title: &Title
        product_description: &Description,
```

File: src/search-filter/src/service/enhanced_search_match_service.rs [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+    ) -> Result<EnhancedSearchMatchResult, EnhancedSearchMatchError>;
+}
+
+pub struct EnhancedSearchMatchServiceImpl {
+    llm: Box<dyn llm::LLMProvider>,
+}
+
+impl EnhancedSearchMatchServiceImpl {
+    pub fn new(api_key: &str) -> Self {
+        let llm = llm::builder::LLMBuilder::new()
```
Comment (bruderj15) [inline src/search-filter/src/service/enhanced_search_match_service.rs:45]: add in retries and co via the builder. check the other usages of `LLMBuilder` in the repository

**Review by bruderj15 — COMMENTED** [review]
@copilot good work. Check the review comments.
